c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,522 @@
+use crate::rewrite::Rewrite;
+use rustc_hir::Mutability;
+use rustc_span::source_map::{FileName, SourceMap};
+use rustc_span::{BytePos, SourceFile, Span, SyntaxContext};
+use std::cmp::Reverse;
+use std::collections::HashMap;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum RewriteError<S = Span> {
+    /// The provided rewrite overlaps, but is not contained in, another rewrite.  `.0` is the span
+    /// of the other rewrite.
+    PartialOverlap(S),
+    /// The provided rewrite conflicts with a different rewrite at the same span.
+    Conflict,
+    /// The provided rewrite affects code that would be discarded by a rewrite of a containing
+    /// expression.  `.0` is the span of the containing expression, and `.1` is its rewrite.
+    _Discarded(S, Box<Rewrite<S>>),
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+struct RewriteTree<S = Span> {
+    span: S,
+    rw: Rewrite<S>,
+    /// Child nodes, representing rewrites that are contained entirely within the current rewrite.
+    /// Each child's `span` is contained within `self.span`, no child's `span` overlaps any other
+    /// child's `span`, and children are sorted by position (`span.lo()`).
+    children: Vec<RewriteTree<S>>,
+}
+
+/// This trait defines the subset of the [`Span`] API that we use in this module.  It's implemented
+/// for `Span` and also for `FakeSpan`, which is a simple span type we use in tests to avoid
+/// dealing with interner/`SourceMap` machinery.
+trait SpanLike: Copy + Eq {
+    fn lo(self) -> BytePos;
+    fn hi(self) -> BytePos;
+    fn contains(self, other: Self) -> bool;
+    fn overlaps(self, other: Self) -> bool;
+}
+
+impl SpanLike for Span {
+    fn lo(self) -> BytePos {
+        self.lo()
+    }
+    fn hi(self) -> BytePos {
+        self.hi()
+    }
+    fn contains(self, other: Self) -> bool {
+        self.contains(other)
+    }
+    fn overlaps(self, other: Self) -> bool {
+        self.overlaps(other)
+    }
+}
+
+impl<S: SpanLike> RewriteTree<S> {
+    pub fn build(
+        mut rws: Vec<(S, Rewrite<S>)>,
+    ) -> (Vec<RewriteTree<S>>, Vec<(S, Rewrite<S>, RewriteError<S>)>) {
+        // Sort by start position and then by decreasing length, so that each parent span comes
+        // before all its children.
+        rws.sort_by_key(|&(ref s, _)| (s.lo(), Reverse(s.hi() - s.lo())));
+
+        // The stack contains a `RewriteTree` node for the most recently processed item from `rws`
+        // along with all of its ancestors (nodes whose spans strictly contain the item's span).
+        // `stack[0]` is the outermost ancestor, and `stack.last()` is the most recently pushed
```

Comment:
I think the relationship between the order of the trees in the stack and the order of `rws` can be made clearer. From a first read it was not obvious that the stack order implies that index `n` is a parent of index `n-1` if there is complete overlap, and the reverse for `rws`. That's made clearer in the code below with things like `if let Some(parent) = stack.last_mut()`

Reply:
I edited the comments in `RewriteTree::build` - let me know if the new text seems clearer.